mitiq/shadows/classical_postprocessing.py

```diff
@@ -8,130 +8,380 @@
+            calibration.
+        f_est: The estimated Pauli fidelity for each calibration
+
+    Returns:
+        Reconstructed classical snapshot in terms of nparray.
+    """
+    # calibrate the noisy quantum channel, output in PTM rep.
+    if pauli_twirling_calibration:
+        if f_est is None:
+            raise ValueError(
+                "estimation of Pauli fidelity must be provided for Pauli"
+                "twirling calibration."
+            )
+        elements = []
+        # get b_list and f for each calibration measurement
+        for b_list_cal, f in f_est.items():
+            pi_snapshot_vecter = []
+            for b_1, b2, u2 in zip(b_list_cal, b_list_shadow, u_list_shadow):
+                # get pi for each qubit based on calibration measurement
+                pi = pi_zero if b_1 == "0" else pi_one
+                # get state for each qubit based on shadow measurement
+                state = ZERO_STATE if b2 == "0" else ONE_STATE
+                # get U2 for each qubit based on shadow measurement
+                U2 = PAULI_MAP[u2]
+                pi_snapshot_vecter.append(
+                    pi * operator_ptm_vector_rep(U2.conj().T @ state @ U2)
+                )
+                # solve for the snapshot state
+            elements.append(1 / f * kronecker_product(pi_snapshot_vecter))
+        rho_snapshot_vector = np.sum(elements, axis=0)
+        # normalize the snapshot state
+        rho_snapshot = rho_snapshot_vector  # * normalize_factor
+    # w/o calibration, noted here, the output in terms of matrix,
+    # not in PTM rep.
+    else:
+        local_rhos = []
+        for b, u in zip(b_list_shadow, u_list_shadow):
+            state = ZERO_STATE if b == "0" else ONE_STATE
+            U = PAULI_MAP[u]
+            # apply inverse of the quantum channel,get PTM vector rep
+            local_rho = 3.0 * (U.conj().T @ state @ U) - cirq.I._unitary_()
+            local_rhos.append(local_rho)
+
+        rho_snapshot = kronecker_product(local_rhos)
     return rho_snapshot
 
 
 def shadow_state_reconstruction(
-    measurement_outcomes: Tuple[NDArray[Any], NDArray[np.string_]]
+    shadow_measurement_outcomes: Tuple[List[str], List[str]],
+    pauli_twirling_calibration: bool,
+    f_est: Optional[Dict[str, float]] = None,
 ) -> NDArray[Any]:
     """Reconstruct a state approximation as an average over all snapshots.
 
     Args:
-        measurement_outcomes: Tuple obtained from ``random_pauli_measurement``.
-
+        shadow_measurement_outcomes: Measurement result and the basis
+            performing the measurement obtained from
+            `random_pauli_measurement` for classical shadow protocol.
+        pauli_twirling_calibration: Whether to use Pauli twirling
+            calibration.
+        f_est: The estimated Pauli fidelity for each calibration
     Returns:
-        The reconstructed quantum state.
+        Numpy array with the reconstructed state.
```

Comment:
Hi Andrea, I'm not sure whether this is a good phrase.  Since the reconstructed state is not a quantum state (it is the outcome of a non-CP channel).  So I'm not sure whether it could be called a density matrix, which itself has to be a positive element of the predual of $\mathcal{B}(\mathcal{H})$ i.e. map positive operator to positive operator.  So cautiously, I'll call it matrix. How is this sounds to you?  Will change it to:
`The state reconstructed from classical shadow protocal`

Reply:
sounds good